### DISCUSSION TITLE
Why resume the consumer before invoke the listeners?

### DISCUSSION THREAD

**Discussion by pengwan75** [spring-projects/spring-kafka#1819] [Q&A] [answered]
Hi there, 

i am looking for the help to understand the logic in the following code block, 

```java 
protected void pollAndInvoke() {
	...
	pauseConsumerIfNecessary();
	...
	resumeConsumerIfNeccessary();
	...
	invokeIfHaveRecords(records);
}
```

what is the reason/benefit to resume the consumer before invoke the listeners?
whether the client heartbeat will occur the pollTimeoutExpired if the "custom" listener take a long time?

Thanks in advance.

**Comment by garyrussell** [comment, marked as answer]
If the consumer is paused (at the application's request), the poll will return no records, so the listener won't be invoked, so it doesn't matter where the resume is done.

An application might pause the container (and hence the consumer before the next poll), if it is not ready to receive new records, but it wants to remain part of the consumer group and not cause a rebalance.

The listener CANNOT "take a long time" - it must be able to handle the records returned by the poll within the configured `max.poll.interval.ms`. Otherwise, the broker will force a rebalance.

pause/resume is simply a mechanism to remain part of the group while not actively receiving any records, but we must continue to poll.
